pkg/cmd/pulumi/state/state_protect.go

```diff
@@ -118,25 +120,62 @@ func protectAllResources(ctx context.Context, ws pkgWorkspace.Context, stackName
 	return nil
 }
 
+// protectResourcesInSnapshot handles the logic for protecting resources in a snapshot.
+func protectResourcesInSnapshot(snap *deploy.Snapshot, urns []string) (int, []error) {
+	if snap == nil {
+		return 0, []error{errors.New("no resources found to protect")}
+	}
+
+	var errs []error
+	resourceCount := 0
+
+	// Map URNs to resources for efficient lookup
+	urnToResource := make(map[resource.URN]*resource.State)
+	for _, res := range snap.Resources {
+		urnToResource[res.URN] = res
+	}
+
+	for _, urnStr := range urns {
+		urn := resource.URN(urnStr)
+		res, found := urnToResource[urn]
+
+		if found {
+			res.Protect = true
+			resourceCount++
+		} else {
+			errs = append(errs, fmt.Errorf("No such resource %q exists in the current state", urn))
+		}
+	}
+
+	return resourceCount, errs
+}
+
+// protectMultipleResources protects multiple resources specified by their URNs.
+func protectMultipleResources(
+	ctx context.Context, ws pkgWorkspace.Context, stackName string, urns []string, showPrompt bool,
+) error {
+	return runTotalStateEditWithPrompt(
+		ctx, ws, backend.DefaultLoginManager, stackName, showPrompt, func(_ display.Options, snap *deploy.Snapshot) error {
+			resourceCount, errs := protectResourcesInSnapshot(snap, urns)
+
+			if resourceCount > 0 && len(errs) == 0 {
+				fmt.Printf("%d resources protected\n", resourceCount)
+			}
+
+			if len(errs) > 0 {
+				var errMsgs []string
+				for _, err := range errs {
+					errMsgs = append(errMsgs, err.Error())
+				}
+				return errors.New(strings.Join(errMsgs, "\n"))
+			}
+
+			return nil
+		}, protectMessage)
+}
+
 func protectResource(
 	ctx context.Context, ws pkgWorkspace.Context, stackName string, urn resource.URN, showPrompt bool,
 ) error {
-	err := runStateEditWithPrompt(
-		ctx,
-		ws,
-		backend.DefaultLoginManager,
-		stackName,
-		showPrompt,
-		urn,
-		func(_ *deploy.Snapshot, res *resource.State) error {
-			res.Protect = true
-			return nil
-		},
-		protectMessage,
-	)
-	if err != nil {
-		return err
-	}
-	fmt.Println("Resource protected")
-	return nil
+	return protectMultipleResources(ctx, ws, stackName, []string{string(urn)}, showPrompt)
 }
```

Comment:
I would have removed `protectResource` here and just called `protectMultipleResources(ctx, ws, stackName, []string{string(urn)}, showPrompt)` on line 89 directly, but I don't mind the alias.